File: test_observatory_timeline_paths.py
"""Regression coverage for all per-event Observatory timeline paths."""
from __future__ import annotations

import copy
import sys
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parent
sys.path.insert(0, str(ROOT))

from protocol.projector import project

NOW = "2026-09-06T23:00:00Z"


def event(event_id: str, paths: list[str], **extra: object) -> dict:
    return {
        "event_id": event_id,
        "kind": "CHECKPOINT",
        "ts": NOW,
        "session_id": "timeline-path-regression",
        "task_id": "multi-file-work",
        "claimed_paths": paths,
        **extra,
    }


class TimelinePathsTest(unittest.TestCase):
    def test_all_paths_are_retained_with_legacy_first_path(self):
        paths = ["src/first.py", "tests/secondary.py", "docs/third.md"]
        row = project([event("paths-multi", paths)], now=NOW)["timeline"][0]
        self.assertEqual(row.get("claimed_paths"), paths)
        self.assertEqual(row["path"], paths[0])

    def test_single_path_remains_compatible(self):
        row = project([event("paths-single", ["src/first.py"])], now=NOW)["timeline"][0]
        self.assertEqual(row.get("claimed_paths"), ["src/first.py"])
        self.assertEqual(row["path"], "src/first.py")

    def test_empty_paths_preserve_unknown_alias(self):
        row = project([event("paths-empty", [])], now=NOW)["timeline"][0]
        self.assertEqual(row.get("claimed_paths"), [])
        self.assertEqual(row["path"], "UNKNOWN")

    def test_paths_belong_to_each_event_not_latest_task_state(self):
        rows = project([
            event("paths-earlier", ["old/a.py", "old/b.py"], ts="2026-09-06T22:58:00Z"),
            event("paths-later", ["new/c.py", "new/d.py"]),
        ], now=NOW)["timeline"]
        by_id = {row["event_id"]: row for row in rows}
        self.assertEqual(by_id["paths-earlier"].get("claimed_paths"), ["old/a.py", "old/b.py"])
        self.assertEqual(by_id["paths-later"].get("claimed_paths"), ["new/c.py", "new/d.py"])

    def test_timeline_paths_are_not_shared_with_other_projections(self):
        source = event("paths-independent", ["src/a.py", "tests/b.py"])
        snap = project([source], now=NOW)
        row = snap["timeline"][0]
        self.assertIsInstance(row.get("claimed_paths"), list)
        row["claimed_paths"].append("not-a-real-event-path")
        self.assertEqual(source["claimed_paths"], ["src/a.py", "tests/b.py"])
        self.assertNotIn("not-a-real-event-path", snap["sessions"][0]["claimed_paths"])
        self.assertNotIn("not-a-real-event-path", snap["work_map"][0]["claimed_paths"])

    def test_projection_is_deterministic_and_does_not_mutate_events(self):
        source = [event("paths-stable", ["src/a.py", "tests/b.py"])]
        original = copy.deepcopy(source)
        self.assertEqual(project(source, now=NOW), project(source, now=NOW))
        self.assertEqual(source, original)


if __name__ == "__main__":
    unittest.main()

File: test_observatory_timeline_paths.js
/* Run the actual renderer in a small DOM harness; no browser/network dependency. */
"use strict";
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");
const vm = require("node:vm");

class Element {
  constructor(tag) {
    this.tagName = tag;
    this.children = [];
    this.listeners = {};
    this._text = "";
  }
  set textContent(value) { this._text = String(value); this.children = []; }
  get textContent() { return this._text + this.children.map(child => child.textContent).join(""); }
  set innerHTML(value) { this.textContent = value; }
  appendChild(child) { this.children.push(child); return child; }
  addEventListener(name, callback) { this.listeners[name] = callback; }
}

async function main() {
  const ids = ["snapshot-status", "refresh-snapshot", "refresh-status", "coverage-note",
    "source-coverage", "cockpit-lines", "metrics", "board-motion", "session-rows",
    "presence-rows", "work-rows", "collision-list", "attention-list", "timeline-filters",
    "timeline", "briefing", "handoff", "economy", "routes"];
  const nodes = Object.fromEntries(ids.map(id => [id, new Element(id)]));
  const row = (id, extra) => ({ts: "2026-09-06T23:00:00Z", kind: "CHECKPOINT",
    session_id: id, task_id: "task-" + id, evidence: [], ...extra});
  let fixture = {
    schema: "commons-observatory/v0.1", now: "2026-09-06T23:00:00Z", stale_after_seconds: 600,
    timeline: [
      row("multi", {path: "src/first.py", claimed_paths: ["src/first.py", "tests/secondary.py", "docs/third.md"]}),
      row("legacy", {path: "legacy/only.py"}),
      row("empty", {path: "UNKNOWN", claimed_paths: []}),
      row("fallback", {path: "fallback/only.py", claimed_paths: null}),
    ],
  };
  let requests = 0;
  let failure = null;
  const pendingTimers = new Map();
  let timerId = 0;
  const context = vm.createContext({
    console, Date, AbortController,
    document: {
      hidden: false,
      getElementById: id => nodes[id],
      createElement: tag => new Element(tag),
      addEventListener() {},
    },
    window: {},
    FormData: class { constructor(values) { this.values = values; } get(key) { return this.values[key] || ""; } },
    setTimeout: (callback, delay) => { pendingTimers.set(++timerId, {callback, delay}); return timerId; },
    clearTimeout: id => pendingTimers.delete(id),
    setInterval() {},
    fetch: async () => {
      requests++;
      if (failure) throw failure;
      return {ok: true, json: async () => fixture};
    },
  });
  const source = fs.readFileSync(path.join(__dirname, "observatory.js"), "utf8");
  vm.runInContext(source, context, {filename: "observatory.js"});
  const settle = () => new Promise(resolve => setImmediate(resolve));
  await settle();
  assert.equal(context.window.COMMONS_OBSERVATORY, fixture, "real renderer should load fixture");
  const matching = () => nodes.timeline.children.filter(li => li.className !== "empty").map(li => {
    const match = /session ([^ ]+) · task/.exec(li.textContent);
    assert.ok(match, "timeline row should contain its session");
    return match[1];
  });
  let checks = 0;
  function select(values, expected, message) {
    nodes["timeline-filters"].listeners.submit({preventDefault() {}, target: values});
    assert.deepEqual(matching(), expected, message);
    checks++;
  }
  select({path: "tests/secondary.py"}, ["multi"], "secondary path should match");
  select({path: "src/first.py"}, ["multi"], "first path remains searchable");
  select({path: "SECONDARY"}, ["multi"], "path search is case-insensitive");
  select({path: "docs/third"}, ["multi"], "third path supports substring search");
  select({path: "missing/file.py"}, [], "unrelated path must not match");
  select({path: "first.py tests"}, [], "do not match across joined filenames");
  select({path: "first.py,tests"}, [], "do not match across comma-joined filenames");
  select({path: "legacy/only"}, ["legacy"], "legacy single-path bake remains readable");
  select({path: "fallback/only"}, ["fallback"], "null list falls back to legacy path");
  select({path: "UNKNOWN"}, ["empty"], "empty path list preserves UNKNOWN behavior");
  select({path: "secondary", kind: "LANDING"}, [], "other filters still combine with paths");
  select({path: "secondary", kind: "CHECKPOINT", session: "multi"}, ["multi"], "combined matching filters retain row");
  nodes["timeline-filters"].listeners.reset();
  for (const [id, timer] of pendingTimers) {
    if (timer.delay === 0) { pendingTimers.delete(id); timer.callback(); }
  }
  assert.deepEqual(matching(), ["multi", "legacy", "empty", "fallback"], "reset restores all rows");
  checks++;
  select({path: "secondary"}, ["multi"], "select before refresh");
  fixture = {...fixture, timeline: [row("refreshed", {path: "src/new.py", claimed_paths: ["src/new.py", "tests/secondary.py"]})]};
  await nodes["refresh-snapshot"].listeners.click();
  assert.deepEqual(matching(), ["refreshed"], "refresh preserves active multi-path filter");
  checks++;
  failure = new Error("fixture network failure");
  await nodes["refresh-snapshot"].listeners.click();
  assert.deepEqual(matching(), ["refreshed"], "failed refresh preserves last successful data and filter");
  assert.equal(nodes["refresh-snapshot"].disabled, false, "failed refresh re-enables the button");
  checks++;
  assert.equal(requests, 3, "filtering and reset must not create network requests");
  checks++;
  console.log(`PASS ${checks} timeline path renderer checks`);
}

main().catch(error => { console.error(error); process.exitCode = 1; });
